code/ship/shiphit.cpp

```diff
@@ -2386,6 +2386,27 @@ static void ship_do_damage(object *ship_objp, object *other_obj, vec3d *hitpos,
 				damage = Armor_types[shipp->armor_type_idx].GetDamage(damage, damage_type_idx, difficulty_scale_factor, other_obj_is_beam);
 				apply_diff_scale = false;
 			}
+		}
+			// if weapon is vampiric, slap healing on shooter instead of target
+		if (Weapons[other_obj->instance].weapon_info_index >= 0){
```

Comment:
This is where you'd use `shiphit_get_damage_weapon(other_obj)` to get the proper `weapon_info_index` before continuing with the rest of the code if you want compatibility with shockwaves and beams.